services/src/floodlight-modules/src/main/java/org/openkilda/floodlight/kafka/producer/OrderAwareWorker.java

```diff
@@ -0,0 +1,84 @@
+/* Copyright 2018 Telstra Open Source
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.openkilda.floodlight.kafka.producer;
+
+import org.apache.kafka.clients.producer.Callback;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+
+import java.util.concurrent.Future;
+
+public class OrderAwareWorker extends AbstractWorker {
+    private int deep = 1;
+    private long expireAt = 0;
+    private Integer partition;
+
+    public OrderAwareWorker(AbstractWorker worker) {
+        super(worker.getKafkaProducer(), worker.getTopic());
+
+        if (worker instanceof OrderAwareWorker) {
+            OrderAwareWorker other = (OrderAwareWorker) worker;
+            this.deep += other.deep;
+            this.partition = other.partition;
+        }
+    }
+
+    public OrderAwareWorker(Producer<String, String> kafkaProducer, String topic) {
+        super(kafkaProducer, topic);
+    }
+
+    @Override
+    protected synchronized SendStatus send(String payload, Callback callback) {
+        ProducerRecord<String, String> record;
+        if (partition == null) {
+            record = new ProducerRecord<>(getTopic(), payload);
+        } else {
+            record = new ProducerRecord<>(getTopic(), partition, null, payload);
+        }
+
+        Future<RecordMetadata> promise = getKafkaProducer().send(record, callback);
+        if (partition == null) {
+            try {
+                partition = promise.get().partition();
+            } catch (Exception e) {
+                // exceptions will be processes by caller on it's own promise.get() call
```

Comment:
anyway, add log error

Reply:
This code is on "library" side, it don't know what to do with errors.

If/when caller call org.openkilda.floodlight.kafka.producer.SendStatus#waitTillComplete it will receive same exception and can handle it in any suitable for him way.

If caller use method org.openkilda.floodlight.kafka.producer.Producer#sendMessageAndTrack, this error will be logged into org.openkilda.floodlight.kafka.producer.Producer#reportError (You can see it here org.openkilda.floodlight.kafka.producer.ProducerTest#errorReporting).

Do you insist on log message in this point?